vim-patch:9.0.{1585,1595}

luukvbaal opened the pull request:
**vim-patch:9.0.1585: weird use of static variables for spell checking**


Problem:    Weird use of static variables for spell checking.
Solution:   Move the variables to a structure and pass them from win_update()
            to win_line(). (Luuk van Baal, closes vim/vim#12448)

https://github.com/vim/vim/commit/30805a1aba0067cf0087f9a0e5c184562433e2e7

**vim-patch:9.0.1595: line pointer becomes invalid when using spell checking**

Problem:    Line pointer becomes invalid when using spell checking.
Solution:   Call ml_get() at the right places. (Luuk van Baal, closes vim/vim#12456)

https://github.com/vim/vim/commit/e84c773d42e8b6ef0f8ae9b6c7312e0fd47909af

luukvbaal committed:
vim-patch:9.0.1585: weird use of static variables for spell checking

Problem:    Weird use of static variables for spell checking.
Solution:   Move the variables to a structure and pass them from win_update()
            to win_line(). (Luuk van Baal, closes vim/vim#12448)

https://github.com/vim/vim/commit/30805a1aba0067cf0087f9a0e5c184562433e2e7

zeertzjq commented on src/nvim/drawline.c:
```diff
@@ -1021,21 +1021,25 @@ static void win_line_continue(winlinevars_T *wlv)
-/// Display line "lnum" of window 'wp' on the screen.
+/// Display line "lnum" of window "wp" on the screen.
+/// Start at row "startrow", stop when "endrow" is reached.
+/// When "number_only" is TRUE only update the number column.
```
```suggestion
/// When "number_only" is true only update the number column.
```

zeertzjq reviewed (changes requested):
You need to pass a zero `spellvars_T` on folded lines and end filler lines. I'm surprised that there are no compiler or clang-tidy warnings for passing `false` as a pointer.

luukvbaal commented:
Quite surprising indeed.

zeertzjq commented:
When I run `test/functional/ui/fold_spec.lua` tests locally on this PR the test `folded lines without ext_multigrid work with spell` fails, but it passes on CI. Strange.

luukvbaal commented:
Right, guess I'll make another patch to vim first to fix this.

zeertzjq commented:
I think it works fine in Vim.

luukvbaal commented:
`line` is cleared by an `ml_get_buf()` call that this patch added.
I think it should be restored by calling `ml_get_buf()` again.

zeertzjq commented:
Oh, `test_override('alloc_lines', 1)` isn't called in Vim instances run with `RunVimInTerminal()`.

luukvbaal committed:
vim-patch:9.0.1595: line pointer becomes invalid when using spell checking

Problem:    Line pointer becomes invalid when using spell checking.
Solution:   Call ml_get() at the right places. (Luuk van Baal, closes vim/vim#12456)

https://github.com/vim/vim/commit/e84c773d42e8b6ef0f8ae9b6c7312e0fd47909af

luukvbaal commented on src/nvim/drawline.c:
```diff
@@ -1021,21 +1021,25 @@ static void win_line_continue(winlinevars_T *wlv)
+/// Start at row "startrow", stop when "endrow" is reached.
+/// When "number_only" is true only update the number column.
+/// "spv" is used to store information for spell checking, kept between
+/// sequential calls for the same window.
```
Is this desired? I noticed it was missing but I guess it is included in short-form in the @param descriptions.

zeertzjq replied:
What was missing?

luukvbaal replied:
(Everything added here?) I mean I noticed that the function comment for our `win_line()` was much shorter. But I guess the information is included in the param comments instead.
Do we want both or should I remove it again?

zeertzjq replied:
I think they should be moved to the param comments.

luukvbaal commented on src/nvim/drawline.c:
```diff
@@ -1420,18 +1377,46 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, int mod_top, bo
 
-  line = end_fill ? "" : ml_get_buf(wp->w_buffer, lnum, false);
-  ptr = line;
+  if (spv->spv_has_spell && !number_only) {
```
`number_only`, `end_fill` and `has_fold`, are used quite inconsistently throughout `win_line()`. I think a lot of times if one is used to guard a block, the others should be as well. That is not always the case currently.

Might verify that this doesn't break anything and/or introduce a new boolean to skip these blocks later...

zeertzjq replied:
In other places this seems already guarded by other things like `v > 0`.

luukvbaal replied:
Why not just add `&& !has_fold && !end_fill` to the spell check block?

zeertzjq replied:
It matches the old code better and make it less necessary to decide whether to add `&& !has_fold && !end_fill` in other places where `spv_has_spell` appears.

zeertzjq committed:
refactor(drawscreen): avoid spell check on folded or filler lines
